step_counter/src/index.ts

```diff
@@ -0,0 +1,129 @@
+// This is one of three related extension examples that demonstrate
+// JupyterLab's provider-consumer pattern, where plugins can depend
+// on and reuse features from one another. The three packages that
+// make up the complete example are:
+//
+//   1. (*) The step_counter package (this one). This holds a token, a
+//      class + an interface that make up a stock implementation of
+//      the "step_counter" service, and a provider plugin that
+//      makes an instance of the Counter available to JupyterLab
+//      as a service object.
+//   2. The step_counter_extension package, that holds a UI/interface
+//      in JupyterLab for users to count their steps that connects
+//      with/consumes the step_counter service object via a consumer plugin.
+//   3. The leap_counter_extension package, that holds an alternate
+//      way for users to count steps (a leap is 5 steps). Like the step_counter_extension
+//      package, this holds a UI/interface in JupyterLab, and a consumer
+//      plugin that also requests/consumes the step_counter service
+//      object. The leap_counter_extension package demonstrates how
+//      an unrelated plugin can depend on and reuse features from
+//      an existing plugin. Users can install either the
+//      step_counter_extension, the leap_counter_extension or both
+//      to get whichever features they prefer (with both reusing
+//      the step_counter service object).
+
+import {
+  JupyterFrontEnd,
+  JupyterFrontEndPlugin
+} from '@jupyterlab/application';
+
+import { Token } from '@lumino/coreutils';
+import { Signal } from '@lumino/signaling';
+
+// The StepCounterItem interface is used as part of JupyterLab's
+// provider-consumer pattern. This interface is supplied to the
+// token instance (the StepCounter token), and JupyterLab will
+// use it to type-check any service-object associated with the
+// token that a provider plugin supplies to check that it conforms
+// to the interface.
+interface StepCounterItem {
```

Comment:
Oh, so I see why you used `any` earlier. The problem here is twofold:
a) `StepCounterItem` is not exported
b) It is named differently from `StepCounter` so we do not get the benefit of [declaration merging](https://www.typescriptlang.org/docs/handbook/declaration-merging.html)

Typically we export just one symbol which is a merge token and interface. I see a potential for didactic value of not doing so, but then in real life we do it because it allows for a clean one symbol export.

Reply:
I purposefully did not use declaration merging because this is a beginner tutorial for these concepts, I think it's VERY likely to cause problems/confusion for non-veterans (again, our student or biologist, chemist, finance users who are smart but probably not experienced web developers), and as a 6-7 year desktop applications developer I can say personally that it caused me some difficulties (and I found it very peculiar and unintuitive language feature/pattern).